### ISSUE TITLE
Allow non-JSON error response to be parsed and request retried

### ISSUE THREAD

**Issue by jkwlui** [googleapis/nodejs-common#583]
It has become known that the service API can return non-JSON response body for some retryable errors like 503 - Service Unavailable.

Here is a stack trace of customer reported failures:
```
err: {
   message: "Cannot parse response as JSON: Service Unavailable"    
   response: {…}    
   stack: "Error: Cannot parse response as JSON: Service Unavailable
    at Util.parseHttpRespBody (/usr/src/app/node_modules/@google-cloud/common/build/src/util.js:189:42)
    at Util.handleResp (/usr/src/app/node_modules/@google-cloud/common/build/src/util.js:135:117)
    at /usr/src/app/node_modules/@google-cloud/common/build/src/util.js:434:22
    at onResponse (/usr/src/app/node_modules/retry-request/index.js:206:7)
    at /usr/src/app/node_modules/teeny-request/build/src/index.js:170:17
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)"    
   type: "ApiError"    
  }
```

We should update `parseHttpRespBody` to allow non-JSON responses:
https://github.com/googleapis/nodejs-common/blob/master/src/util.ts#L413-L434

**Comment by stephenplusplus** [comment]
I believe retrying is determined by the status code of the response. Even if the error body is not JSON, I would think the status code is whatever is appropriate for the given message. Of course, I'd have to confirm-- do we know how to reproduce this error? Any more information?

**Comment by JustinBeckwith** [comment]
@jkwlui is this really a P1?  Is there a customer bug that you can reference that outlines where this is causing trouble?

**Comment by jkwlui** [comment]
ref: internal bug 157979420

**Comment by JustinBeckwith** [comment]
Awesome.  @jkwlui you want to take a whack at this?

**Comment by pebo** [comment]
This issue appears a couple of times every day in production when in writing to GCS from cloud run using the node storage library. 

The GCS backend servers occasionally seem to response with a non-json payload and the nodejs-common library  then fails parsing the response body as it assumes content-type application/json without looking at the response headers.

Note that the GCS backend service sometimes responds with `text/plain` such as "Service Unavailable" but it can also respond with `text/html`, resulting in errors like:

```
"Error: Cannot parse response as JSON: <!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 502 (Server Error)!!1</title>
  <style>
    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>502.</b> <ins>That’s an error.</ins>
  <p>The server encountered a temporary error and could not complete your request.<p>Please try again in 30 seconds.  <ins>That’s all we know.</ins>

    at Util.parseHttpRespBody (/usr/src/app/node_modules/@google-cloud/common/build/src/util.js:189:42)
    at Util.handleResp (/usr/src/app/node_modules/@google-cloud/common/build/src/util.js:135:117)
    at /usr/src/app/node_modules/@google-cloud/common/build/src/util.js:434:22
    at onResponse (/usr/src/app/node_modules/retry-request/index.js:206:7)
    at /usr/src/app/node_modules/teeny-request/build/src/index.js:170:17
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)"
```

**Comment by stephenplusplus** [comment]
Interesting, thanks @pebo. Per https://github.com/googleapis/nodejs-common/blob/ca3f0b381e3361055c96c1e51ba0fbe26661c7c1/src/util.ts#L422, we're just taking the non-JSON response body and putting it as the message of the error. I believe the claim of this issue is that when we get a non-JSON response on a request that should be retried, we do not. But, the HTTP response should still come with a status code that we use to determine if we want to retry. I can't reproduce this scenario to verify that, however.

Should we consider the "Cannot parse response as JSON" error message as being a different bug, where we'd just want to clean that up to use the generic status message for the given code? So in the case above, instead of the HTML error, we'd just use "Error: Bad Gateway".

**Comment by pebo** [comment]
Thanks @stephenplusplus, the stack trace I added was from an application using @google-cloud/storage v5.1.1 and an older version of util.ts than the current one. The fix at L422 of util.ts was added in https://github.com/googleapis/nodejs-common/commit/dbaad170a2e3a6785568523086f15d88fb34eaca (google-cloud/common v3.3.2?) and this version is included in the latest version of  @google-cloud/storage as far as I can tell
````
├─┬ @google-cloud/storage@5.2.0
│ ├─┬ @google-cloud/common@3.3.2
````
Hence I think that the issue with the somewhat misleading error message "Cannot parse response as JSON" can be considered resolved. I've deployed new versions of our services with storage@5.2.0 and will monitor the logs.

Responses with HTTP status code 502, 503 etc are probably being retried:
https://github.com/googleapis/nodejs-common/blob/dbaad170a2e3a6785568523086f15d88fb34eaca/src/util.ts#L735 https://github.com/googleapis/nodejs-common/blob/dbaad170a2e3a6785568523086f15d88fb34eaca/src/util.ts#L515 
but I guess that we'll still see these errors surfacing if all retries attempts fail?
